synapse/storage/state.py

```diff
@@ -785,6 +790,23 @@ def _get_state_for_groups(
             groups, state_filter or StateFilter.all()
         )
 
+    async def _get_state_group_for_events(
+        self,
+        event_ids: Collection[str],
+        await_full_state: bool = True,
+    ) -> Mapping[str, int]:
+        """Returns mapping event_id -> state_group
+
+        Args:
+            event_ids: events to get state groups for
+            await_full_state: if true, will block if we do not yet have complete
+               state at this event.
+        """
+        if await_full_state:
+            await self._partial_state_events_tracker.await_full_state(event_ids)
```

Comment:
That's part of the plan. Another part of the plan is to get cancellation working (https://github.com/matrix-org/synapse/issues/3528) so that we don't stack up requests. A final part of the plan is probably to change the api to do something different, but we're a way off that yet.